File: typescript/packages/aholo-sdk-core/src/poll.ts
import { PollingFailedError, PollingTimeoutError } from './errors.js';

export interface PollOptions {
  /** Poll interval in ms. Default: 500. */
  intervalMs?: number;
  /** Max wait in ms. Default: 300000 (5 min). */
  timeoutMs?: number;
  signal?: AbortSignal;
}

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    if (signal?.aborted) {
      reject(signal.reason ?? new Error('Polling aborted'));
      return;
    }
    const timer = setTimeout(resolve, ms);
    signal?.addEventListener(
      'abort',
      () => {
        clearTimeout(timer);
        reject(signal.reason ?? new Error('Polling aborted'));
      },
      { once: true },
    );
  });
}

export async function pollUntil<T>(options: {
  fn: () => Promise<T>;
  isDone: (result: T) => boolean;
  isFailed?: (result: T) => boolean;
  failMessage?: (result: T) => string;
  poll?: PollOptions;
}): Promise<T> {
  const intervalMs = options.poll?.intervalMs ?? 500;
  const timeoutMs = options.poll?.timeoutMs ?? 300_000;
  const started = Date.now();

  while (true) {
    const result = await options.fn();

    if (options.isFailed?.(result)) {
      throw new PollingFailedError(options.failMessage?.(result) ?? 'Polling failed');
    }
    if (options.isDone(result)) {
      return result;
    }

    if (Date.now() - started >= timeoutMs) {
      throw new PollingTimeoutError(`Polling timed out after ${timeoutMs}ms`, { timeoutMs });
    }

    await sleep(intervalMs, options.poll?.signal);
  }
}

File: python/packages/aholo-sdk-core/src/manycore/aholo_sdk_core/poll.py
from __future__ import annotations

import time
from typing import Callable, Optional, TypeVar

from .errors import PollingFailedError, PollingTimeoutError

T = TypeVar("T")


def poll_until(
    fn: Callable[[], T],
    *,
    is_done: Callable[[T], bool],
    is_failed: Optional[Callable[[T], bool]] = None,
    fail_message: Optional[Callable[[T], str]] = None,
    interval_ms: int = 500,
    timeout_ms: int = 300_000,
) -> T:
    started = time.monotonic()
    while True:
        result = fn()
        if is_failed and is_failed(result):
            raise PollingFailedError(fail_message(result) if fail_message else "Polling failed")
        if is_done(result):
            return result
        if (time.monotonic() - started) * 1000 >= timeout_ms:
            raise PollingTimeoutError(f"Polling timed out after {timeout_ms}ms", timeout_ms=timeout_ms)
        time.sleep(interval_ms / 1000.0)
